kong/resolver/access.lua

```diff
@@ -32,27 +32,93 @@ local function get_host_from_url(val)
   return parsed_url.host..(port and ":"..port or "")
 end
 
--- Retrieve the API from the Host that has been requested
-function _M.execute(conf)
-  local hosts_headers = {}
+-- Find an API from a request made to nginx. Either from one of the Host or X-Host-Override headers
+-- matching the API's `public_dns`, either from the `request_uri` matching the API's `path`.
+--
+-- To perform this, we need to query _ALL_ APIs in memory. It is the only way to compare the `request_uri`
+-- as a regex to the values set in DB. We keep APIs in the database cache for a longer time than usual.
+-- @see https://github.com/Mashape/kong/issues/15 for an improvement on this.
+--
+-- @return `err`         Any error encountered during the retrieval.
+-- @return `api`         The retrieved API, if any.
+-- @return `hosts`       The list of headers values found in Host and X-Host-Override.
+-- @return `request_uri` The URI for this request.
+local function find_api()
+  local retrieved_api
+
+  -- retrieve all APIs
+  local apis_dics, err = cache.get_or_set("APIS_BY_PUBLIC_DNS", function()
+    local apis, err = dao.apis:find_all()
+    if err then
+      return nil, err
+    end
+
+    -- build dictionnaries of public_dns:api and path:apis for efficient lookup.
+    local dns_dic, path_dic = {}, {}
+    for _, api in ipairs(apis) do
+      if api.public_dns then
+        dns_dic[api.public_dns] = api
+      end
+      if api.path then
+        path_dic[api.path] = api
```

Comment:
I'm not sure if LUA has a way to instantiate a Regular Expression, but if it does I would create the regular expressions here so that you don't have to take the performance hit when doing a lookup.